docs/proposals/lora-autoscaler/README.md

```diff
@@ -0,0 +1,352 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The metric is a reactive indicator, which means the latency is unavoidable, the same as HPA. But we'll try to mitigate the
+latency by offering different policies for configuration. The scaling down has the same problem in response to the cost.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### The LoRA Autoscaler
+
+Right now, vLLM has one lora metric `vllm:lora_requests_info` containing three labels:
+
+- running_lora_adapters: a per-adapter count of the number requests running using that adapter, formatted as a comma-separated string
+- waiting_lora_adapters: similar, except counting requests that are waiting to be scheduled
+- max_lora: the static "max number of LoRAs in a single batch." configuration
+
+We will leverage the `waiting_lora_adapters` as the dominant metric for the autoscaling decision.
+
+At a high level, the workflow looks like this:
+
+- Create Playground or Inference Service with the lora configured
+- Dispatch the LoRAs to the instances as much equally as we can, for instance:
+  - if we have 2 replicas with 3 loras, we may dispatch the LoRAs to the replicas as follows:
+    - Replica 1: lora-1, lora-3
+    - Replica 2: lora-2
+  - if we have 7 replicas with 3 loras, we may dispatch the LoRAs to the replicas as follows:
+    - Replica 1: lora-1
+    - Replica 2: lora-2
+    - Replica 3: lora-3
+    - Replica 4: lora-1
+    - Replica 5: lora-2
+    - Replica 6: lora-3
+    - Replica 7: lora-1
+
+  Make sure **at least one lora exists** in replicas, to avoid lora loading overhead in runtime.
+- Once the lora model loaded successfully, the gateway will update the route table for the lora requests
+- The LoRA autoscaler will monitor the `waiting_lora_adapters` metrics:
+
+  - once exceed the target threshold, the **lora autoscaler**, another controller, will jump in. It will trigger the lora loading for the hot loras but not beyond the max_lora configuration and same loras can't be in the same replica which is meaningless.
+  - also once a lora is **under low load**, lora autoscaler will first cut the corresponding traffic to the lora server and then offload the lora model. Note that the offload threshold should be bigger than the loading threshold to avoid the frequent loading/unloading overhead, both of them should be configurable.
+
+Several concerns here about the lora autoscaling:
+
+  - the metric algorithm: right now, we'll use the `waiting_lora_adapters` as the dominant metric for the autoscaling decision
+  - How lora server knows when to load/offload the lora: we'll have a new CRD for tracking the lora loading status
```

Comment:
Can you elaborate on the design details of this part?

Reply:
Will do, it's just a draft.